staging/src/k8s.io/cloud-provider/controllers/route/route_controller.go

```diff
@@ -361,14 +401,39 @@ func (rc *RouteController) isResponsibleForRoute(route *cloudprovider.Route) boo
 	return false
 }
 
-// checks if a node owns a route with a specific cidr
-func hasRoute(rm map[types.NodeName][]*cloudprovider.Route, nodeName types.NodeName, cidr string) bool {
+// checks if a node owns a route with a specific cidr and target Node addresses.
+func hasRoute(rm map[types.NodeName][]*cloudprovider.Route, nodeName types.NodeName, cidr string, realNodeAddrs []v1.NodeAddress) bool {
 	if routes, ok := rm[nodeName]; ok {
 		for _, route := range routes {
 			if route.DestinationCIDR == cidr {
-				return true
+				if len(route.TargetNodeAddresses) == 0 {
+					return true
+				}
+				if haveSameNodeAddrs(realNodeAddrs, route.TargetNodeAddresses) {
+					return true
+				}
+				klog.Warningf("Node addresses have changed from %v to %v", route.TargetNodeAddresses, realNodeAddrs)
 			}
 		}
 	}
 	return false
 }
+
+func haveSameNodeAddrs(addrs0 []v1.NodeAddress, addrs1 []v1.NodeAddress) bool {
+	if len(addrs0) != len(addrs1) {
+		return false
+	}
+	for _, ip0 := range addrs0 {
+		found := false
+		for _, ip1 := range addrs1 {
+			if reflect.DeepEqual(ip0, ip1) {
+				found = true
+				break
+			}
+		}
+		if !found {
```

Comment:
So we are only searching for `addrs0[0]`? we don't care about `addrs[1..]`? If so why do we nested loops?

Reply:
If `found` is false, it returns false early. If `found` is true, then the outer loop will continue: https://github.com/kubernetes/kubernetes/blob/f1dac70e53564762476f5e76b380032d15f019f7/staging/src/k8s.io/cloud-provider/controllers/route/route_controller.go#L429-L431

Reply:
ah. it is equality not searching!. Thanks.